### PR TITLE
Update ERC-7751 revert bubbling

### PR DESCRIPTION
**Pull request by gretzke**

### CODE REVIEW

**Review by marktoda — APPROVED** [review]
looks good to me

File: src/libraries/CustomRevert.sol [hunk 1]
```diff
@@ -75,23 +78,33 @@ library CustomRevert {
         }
     }
 
-    /// @notice bubble up the revert message returned by a call and revert with the selector provided
-    /// @dev this function should only be used with custom errors of the type `CustomError(address target, bytes revertReason)`
-    function bubbleUpAndRevertWith(bytes4 selector, address addr) internal pure {
+    /// @notice bubble up the revert message returned by a call and revert with a wrapped ERC-7751 error
+    function bubbleUpAndRevertWith(address target, bytes4 functionSelector, bytes4 additionalContext) internal pure {
```
Comment (marktoda) [inline src/libraries/CustomRevert.sol:82]: consider renaming these to be a bit more clear, maybe:
address revertingContract
address revertingFunctionSelector
bytes4 additionalContext

Comment (marktoda) [reply src/libraries/CustomRevert.sol:82]: also -- why is `additionalContext` only bytes4 here but bytes in the custom error? seems context could be a custom error w/ params so should we support that here too?

Comment (marktoda) [reply src/libraries/CustomRevert.sol:82]: I guess keeping it bytes4 makes the encoding more simple if we specifically dont plan on having params here

Comment (gretzke) [reply src/libraries/CustomRevert.sol:82]: any additional context we had in the reverts was the target address (hook, erc20 token, recipient) based on the action, so we only needed 4bytes for the selector

**Review by marktoda — APPROVED** [review]

**Review by marktoda — APPROVED** [review]

File: src/types/Currency.sol [hunk 2]
```diff
@@ -30,13 +30,11 @@ function greaterThanOrEqualTo(Currency currency, Currency other) pure returns (b
 library CurrencyLibrary {
     using CustomRevert for bytes4;
```
Comment (hensha256) [inline src/types/Currency.sol:31]: can this line go now?

File: src/libraries/CustomRevert.sol [hunk 3]
```diff
@@ -75,23 +78,43 @@ library CustomRevert {
...
+                and(revertingFunctionSelector, 0xffffffff00000000000000000000000000000000000000000000000000000000)
+            )
+            // offset revert reason
+            mstore(add(fmp, 0x44), 0x80)
+            // offset additional context
+            mstore(add(fmp, 0x64), add(0xa0, encodedDataSize))
+            // size revert reason
+            mstore(add(fmp, 0x84), size)
+            // revert reason
+            returndatacopy(add(fmp, 0xa4), 0, size)
```
Comment (0age) [inline src/libraries/CustomRevert.sol:109]: It's actually cheaper to use returndatasize in place rather than assigning it to a stack variable!

**Review by 0age — DISMISSED** [review]

**Review by 0age — APPROVED** [review]

**Review by hensha256 — APPROVED** [review]

**Pull request merged**